cli/integration-test/integration/client.clj

```diff
@@ -0,0 +1,192 @@
+  IClient
+  (start [this]
+    (reset! listener (future (listen! server-out this))))
+  (shutdown [_this] ;; simulate client closing
+    (.close server-in))
+  (exit [_this] ;; wait for shutdown of server to propagate to listener
+    @@listener)
+  (send-request [this method body]
+    (let [req (lsp-rpc this method body)]
+      (log this :cyan "sending request:" req)
+      (wire-send server-in req)
+      (let [p (promise)]
+        (swap! sent-requests assoc (:id req) p)
+        p)))
+  (send-notification [this method body]
+    (let [notif (lsp-rpc this method body)]
+      (log this :blue "sending notification:" notif)
+      (wire-send server-in notif)))
+  (receive-response [this {:keys [id] :as resp}]
+    (if-let [request (get @sent-requests id)]
+      (do (log this :green "received reponse:" resp)
+          (swap! sent-requests dissoc id)
+          (deliver request (if (:error resp)
+                             resp
+                             (:result resp))))
+      ;; TODO: if we don't have a request, this will return nil, which will get
+      ;; derefed, which will throw an error. Better to use promesa, and return a
+      ;; rejected promise? Promesa has the benefit of using CompletableFuture,
+      ;; making it more like what a lsp4j server expects.
+      (log this :red "received response for unmatched request:" resp)))
+  (receive-request [this {:keys [id method] :as req}]
+    (log this :magenta "received request:" req)
+    (swap! received-requests conj req)
+    (when-let [mock-resp (get @mock-responses (keyword method))]
+      (let [resp {:id id
+                  :result mock-resp}]
+        (log this :magenta "sending mock response:" resp)
+        (wire-send server-in resp))))
+  (receive-notification [this notif]
+    (log this :blue "received notification:" notif)
+    (swap! received-notifications conj notif))
+  IMockClient
+  (mock-response [_this method body]
+    (swap! mock-responses assoc method body)))
+
+(defonce client-id (atom 0))
+
+(defn client [server-in server-out]
+  (map->Client
+    {:client-id (swap! client-id inc)
+     :server-in server-in
+     :server-out server-out
+     :request-id (atom 0)
+     :sent-requests (atom {})
+     :received-requests (atom [])
+     :received-notifications (atom [])
+     :mock-responses (atom {})
+     :listener (atom nil)}))
+
+(defn ^:private keyname [key] (str (namespace key) "/" (name key)))
+
+(defn ^:private await-first-and-remove! [client pred coll-type]
+  (let [coll* (coll-type client)]
+    (loop [tries 0]
+      (if (< tries 20)
+        (if-let [elem (first (filter pred @coll*))]
+          (do
+            (swap! coll* #(->> % (remove #{elem}) vec))
+            elem)
+          (do
+            (Thread/sleep 500)
+            (recur (inc tries))))
+        (do
+          (log client :red "timeout waiting:" coll-type)
+          (throw (ex-info "timeout waiting for client to receive req/notif" {:coll-type coll-type})))))))
```

Comment:
This is new code. It's unrelated to the rest of this PR... I'm tired of waiting the full 10 minutes for the flaky tests to time out. Now, if the server request doesn't arrive within 10 seconds, the tests fail.

Reply:
awesome!
The only concern I have is when tests really take some time to finish, I mean, non graalvm tests usually takes way more time since we start the server multiple times, do you see any chance to affect that?

Reply:
I don't think it'll be a problem... this isn't a timeout for the whole test, or even for the entire lifetime of a server, just for a moment in its lifetime. More specifically we use this helper when we've done something (A) that'll cause the server to send a message to the client asynchronously (B). We do A and then start waiting, so, the timeout applies to the time delta between A and B.

In the sample project, no commands, or analysis, or anything like that should take more than 10 seconds. The one time I can think of where this delta might get anywhere near 10 seconds is startup, which is tested in `integration.initialize-test`. But even there it looks like pretty safe. There the client sends the server a request to initialize. The server creates a bunch of $/progress notifications asynchronously, then responds with its capabilities. The test waits for this response as its first assertion. Its second assertion uses the helper in question to assert that $/progress was sent. Since it should have already been sent, there shouldn't be any wait. Finally, the test sends a notification from the client to the server that the client has been initialized. Then it uses the helper to start waiting for the server to ask the client to register file watchers. I'm not sure whether that file-watcher request happens async during server init, or if it's triggered async when the client says it's finished init. n either case, since we've already waited for the slow part—server init—it shouldn't take 10 more seconds for the file-watcher request to come through.

Anyway, that's a very long description. tldr; I think we're safe.

Reply:
yes, makes sense, thanks for the explanation.
BTW, officially, any request can only be sent after server responded with initialized